### PR TITLE
Added Cooley-Tukey in Common Lisp

### PR DESCRIPTION
**Pull request by 0xJonas**
Essentially a port of my Scala PR (#746) which appears to be stuck in review.

### CODE REVIEW

File: contents/cooley_tukey/code/clisp/fft.lisp [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+  "Actual iterative implementation of the Cooley-Tukey method."
+  (if (>= stride (length data))
+    data
+    (progn
+      (loop for i from 0 below (length data) by (* 2 stride) do
+        (butterfly-group data i stride))
+      (cooley-tukey-tailrec data (* 2 stride)))))
+
+(defun cooley-tukey-iter (data)
+  "Performs the DFT using the iterative Cooley-Tukey method."
```
Comment (Trashtalk217) [inline contents/cooley_tukey/code/clisp/fft.lisp:83]: Are we doing a DFT or an FFT, wording gets somewhat confusing here.

File: contents/cooley_tukey/code/clisp/fft.lisp [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+           (b-index (+ start i stride))
+           (a (elt data a-index))
+           (b (elt data b-index))
+           (coeff (coefficient 1.0 i (* 2 stride))))
+    (multiple-value-bind (sum difference) (butterfly a b coeff)
+      (setf (elt data a-index) sum)
+      (setf (elt data b-index) difference)))))
+
+(defun cooley-tukey-tailrec (data stride)
+  "Actual iterative implementation of the Cooley-Tukey method."
```
Comment (Trashtalk217) [inline contents/cooley_tukey/code/clisp/fft.lisp:74]: I understand what you're trying to say here with 'iterative,' but it's still recursive. Maybe clarify with a comment. Maybe I'm just nitpicking.

File: contents/cooley_tukey/code/clisp/fft.lisp [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+    (loop for i from 0 below (length data)
+      if (evenp i)
+        collect (elt data i) into evens
+      else
+        collect (elt data i) into odds
+      finally
+        (return (merge-sub-ffts (cooley-tukey-rec evens)
+                                (cooley-tukey-rec odds))))))
+
+(defun reverse-bits (value num-bits)
```
Comment (Trashtalk217) [inline contents/cooley_tukey/code/clisp/fft.lisp:38]: A very clever way of reversing bits, good job!

File: contents/cooley_tukey/code/clisp/fft.lisp [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+         (odd-terms (mapcar #'(lambda (v i) (* v (coefficient 1.0 i fft-length)))
+                            odds
+                            indices)))
+    (concatenate 'list 
+                 (mapcar #'+ evens odd-terms)
+                 (mapcar #'- evens odd-terms))))
+
+(defun cooley-tukey-rec (data)
+  "Performs the DFT using the recursive Cooley-Tukey method."
+  (if (<= (length data) 1)
```
Comment (Trashtalk217) [inline contents/cooley_tukey/code/clisp/fft.lisp:27]: The general formatting in use for 'if'-statements is:
```
(if (foo 42)
    (bar 12)
    (baz 103))
```
So add a couple of extra spaces.

File: contents/cooley_tukey/code/clisp/fft.lisp [hunk 5]
```diff
@@ -0,0 +1,103 @@
+
+(defun coefficient (time-index freq-index dft-len)
+  "Calculates a single twiddle factor for the Fourier Transform."
+  (exp (- (/ (* #c( 0 1) 2.0 pi time-index freq-index)
```
Comment (Trashtalk217) [inline contents/cooley_tukey/code/clisp/fft.lisp:4]: One space too many
```suggestion
  (exp (- (/ (* #c(0 1) 2.0 pi time-index freq-index)
```

**Review by Trashtalk217 — CHANGES_REQUESTED** [review]
An excellent submission, glad that you're interested @0xJonas. I've been the single person implementing the algorithms in Common Lisp up until now and I'm happy to see another Lisper in the archive.

One last thing, It would maybe help to put some extra comments throughout the code. The archive tries to provide good code examples for people learning programming languages and it can be helpful to provide some extra information.

The code is still fine as is, however, and if you correct the small formatting problems, it should be ready for merging.

**Comment by 0xJonas** [comment]
- Changed indentation on the if-statements
- Fixed that one whitespace
- Made iterative implementation use a loop. It is actually much simpler now.
- Added comments

File: contents/cooley_tukey/code/clisp/fft.lisp [hunk 6]
```diff
@@ -14,38 +14,46 @@
...
-        collect (elt data i) into evens
-      else
-        collect (elt data i) into odds
-      finally
-        (return (merge-sub-ffts (cooley-tukey-rec evens)
-                                (cooley-tukey-rec odds))))))
+      data
+      (loop
+        for i from 0 below (length data)
+        ;; Split input into even and odd indices into two smaller lists.
```
Comment (Trashtalk217) [inline contents/cooley_tukey/code/clisp/fft.lisp:33]:
```suggestion
        ;; Split even and odd indexed elements into two seperate lists.
```

File: contents/cooley_tukey/code/clisp/fft.lisp [hunk 7]
```diff
@@ -14,38 +14,46 @@
...
-    value
-    (let* ((num-low-bits (floor (/ num-bits 2)))
-           (num-high-bits (- num-bits num-low-bits))
-           (bit-mask (- (expt 2 num-low-bits) 1))
-           (lower-half (logand value bit-mask))
-           (upper-half (ash value (- num-low-bits))))
-      (logior (ash (reverse-bits lower-half num-low-bits) num-high-bits)
-              (reverse-bits upper-half num-high-bits)))))
+      value
+      ;; Split bits into two parts
```
Comment (Trashtalk217) [inline contents/cooley_tukey/code/clisp/fft.lisp:48]:
```suggestion
      ;; Split bits into two parts.
```

File: contents/cooley_tukey/code/clisp/fft.lisp [hunk 8]
```diff
@@ -14,38 +14,46 @@
 (defun merge-sub-ffts (evens odds)
   "Combines the FFTs of the even and odd indices"
```
Comment (Trashtalk217) [inline contents/cooley_tukey/code/clisp/fft.lisp:15]:
```suggestion
  "Combines the FFTs of the even and odd indices."
```

File: contents/cooley_tukey/code/clisp/fft.lisp [hunk 9]
```diff
@@ -14,38 +14,46 @@
...
   (let* ((fft-length (+ (length evens) (length odds)))
-         (indices (loop for i from 0 below (length odds) collect i))
-         (odd-terms (mapcar #'(lambda (v i) (* v (coefficient 1.0 i fft-length)))
-                            odds
-                            indices)))
+         ;; Calculate coefficients for the odd indices.
+         (twiddle-factors (loop for i from 0 below (length odds)
+                             collect (coefficient 1.0 i fft-length)))
+         ;; Multiply values with coefficients
```
Comment (Trashtalk217) [inline contents/cooley_tukey/code/clisp/fft.lisp:20]:
```suggestion
         ;; Multiply values with coefficients.
```

File: contents/cooley_tukey/code/clisp/fft.lisp [hunk 10]
```diff
@@ -14,38 +14,46 @@
...
+         (odd-terms (mapcar #'* odds twiddle-factors)))
+    ;; Combine the two FFTs
```
Comment (Trashtalk217) [inline contents/cooley_tukey/code/clisp/fft.lisp:22]:
```suggestion
    ;; Combine the two FFTs.
```

File: contents/cooley_tukey/code/clisp/fft.lisp [hunk 11]
```diff
@@ -14,38 +14,46 @@
...
+      (loop
+        for i from 0 below (length data)
+        ;; Split input into even and odd indices into two smaller lists.
+        if (evenp i)
+          collect (elt data i) into evens
+        else
+          collect (elt data i) into odds
+        finally
+          ;; Calculate the Fourier Transform for the two smaller lists and
+          ;; combine them into the Fourier Transform for the full input.
```
Comment (Trashtalk217) [inline contents/cooley_tukey/code/clisp/fft.lisp:40]:
```suggestion
          ;; combine them into the Fourier Transform of the full input.
```

**Review by Trashtalk217 — CHANGES_REQUESTED** [review]
Thank you for the quick response, I assumed that you were going to do points for sentences, so I left some suggestions for the comments.

I've rechecked some of the other common lisp code and how comments and docstrings are handled is all over the place (even when one person made all of them), so don't worry about it. Just make sure it's consistent within the file.

**Comment by 0xJonas** [comment]
Oops, I forgot to update a line number after the previous commit.

**Review by Trashtalk217 — APPROVED** [review]

**Comment by Trashtalk217** [comment]
Sorry for the late confirmation, I misunderstood something. But it's good to merge!

**Pull request merged**